examples/dora_finetuning/dora_finetuning.py

```diff
@@ -0,0 +1,173 @@
+import os
+import torch
+from transformers import AutoTokenizer, AutoModelForCausalLM, Trainer, TrainingArguments, DataCollatorWithPadding, BitsAndBytesConfig
+from datasets import load_dataset
+from peft import LoraConfig, get_peft_model
+
+def train_model(
+    base_model: str,
+    data_path: str,
+    output_dir: str,
+    batch_size: int,
+    num_epochs: int,
+    learning_rate: float,
+    cutoff_len: int,
+    val_set_size: int,
+    use_dora: bool,
+    quantize: bool,
+    eval_step: int,
+    save_step: int,
+    device: str,
+    lora_r: int,
+    lora_alpha: int,
+    lora_dropout: float,
+    lora_target_modules: str,
+    hub_model_id: str,
+    push_to_hub: bool,
+):
+    os.environ["TOKENIZERS_PARALLELISM"] = "false"
+    hf_token = os.getenv("HF_TOKEN")
+
+    # Setup device
+    device = torch.device(device)
+    print(f"Using device: {device}")
+    
+    #load tokenizer
+    tokenizer = AutoTokenizer.from_pretrained(base_model, token=hf_token) 
+     
+    #Load model in Quantized and Peft configz  
+    #QDoRA: IF YOU WANNA QUANTIZE THE MODEL
+    if quantize:
+        model = AutoModelForCausalLM.from_pretrained(
```

Comment:
When quantizing, please call `prepare_model_for_kbit_training`, same as in the notebook.

Reply:
updated to reflect `model = prepare_model_for_kbit_training(model, use_gradient_checkpointing=True)`